### PR TITLE
Support in-place update containers with launch priority

### LINKED ISSUE #901: [feature request] Support ordered execution of the application containers during in-place updates in CloneSet
**What would you like to be added**:

We would like CloneSet to support ordered execution of the application containers during in-place updates.

Specifically, 
- user can specify the priority for each application container in a pod. 
- When in-place updates happen  (e.g version change) in the multiple applications containers, the application containers stop and start based on the priority/ordering container by container.
- in-place update to _a_ container (not multiple containers) with specified ordering just have effect to that container, and no impact to the other containers even though they are specified with orders. 
- For containers that are not specified with the orders/priority, no guarantees of the execution order is fine.
- In the case of containers with specified orders fail at kubelet probes and gets restarted by kubelet, the ordering might not apply anymore. But it will be a good-to-have to ensure ordering restarting as well. 
 
**Why is this needed**:

We believe this is a critical feature that can help us to leverage the CloneSet in-place updates successfully.

We have business logic have to be lined up for execution. While we can (and are) leveraging init container to ensure the execution ordering for some business logic. However, updates to the init containers will cause the whole pod to be recreated.
Even though we can do in-place update in the application container, the nature of the init container updates defeat the whole purpose of in-place updates. 

While we can separate the business logic in init container per pod lifecycle to avoid updates in init container, but in practice, there are business logic is not per pod lifecycle but per container lifecycle and we have to line them up for ordered execution. With the feature support, we can put the business logic into long running service as in application container and achieve ordered in-place updates

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

### Ⅰ. Describe what this PR does
Support in-place update containers with launch priority.

- For pods without container orders, no guarantees of the execution order during in-place update multiple containers.
- For pods with container orders:
  - keep execution order during in-place update multiple containers with different priorities.
  - no guarantees of the execution order during in-place update multiple containers with the same priority.

### Ⅱ. Does this pull request fix one issue?

fixes #901

### CODE REVIEW

File: apis/apps/pub/inplace_update.go [hunk 1]
```diff
@@ -61,6 +61,23 @@ type InPlaceUpdateState struct {
 
 	// UpdateEnvFromMetadata indicates there are envs from annotations/labels that should be in-place update.
 	UpdateEnvFromMetadata bool `json:"updateEnvFromMetadata,omitempty"`
+
+	// NextContainerImages is the containers with lower priority that waiting for in-place update in next batch.
+	NextContainerImages map[string]string `json:"nextContainerImages,omitempty"`
+
+	// NextContainerImages is the containers with lower priority that waiting for in-place update in next batch.
+	NextContainerRefMetadata map[string]metav1.ObjectMeta `json:"nextContainerRefMetadata,omitempty"`
```
Comment (furykerry) [inline apis/apps/pub/inplace_update.go:69]: plz correct the comment about NextContainerRefMetadata

Comment (FillZpp) [reply apis/apps/pub/inplace_update.go:69]: updated

File: pkg/util/inplaceupdate/inplace_update_defaults.go [hunk 2]
```diff
@@ -100,55 +230,91 @@ func defaultCalculateInPlaceUpdateSpec(oldRevision, newRevision *apps.Controller
...
 		idx, _ := strconv.Atoi(words[3])
 		if len(oldTemp.Spec.Containers) <= idx {
 			return nil
 		}
-		updateSpec.ContainerImages[oldTemp.Spec.Containers[idx].Name] = jsonPatchOperation.Value.(string)
+		updateSpec.ContainerImages[oldTemp.Spec.Containers[idx].Name] = op.Value.(string)
 	}
-	if metadataChanged {
+
+	if len(metadataPatches) > 0 {
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_defaults.go:266]: no need to check len, iteration through empty map will do nothing

File: pkg/util/inplaceupdate/inplace_update_defaults.go [hunk 3]
```diff
@@ -100,55 +230,91 @@ func defaultCalculateInPlaceUpdateSpec(oldRevision, newRevision *apps.Controller
...
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlaceUpdateEnvFromMetadata) {
+			for _, op := range metadataPatches {
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_defaults.go:268]: plz add a path example here

File: pkg/util/inplaceupdate/inplace_update_test.go [hunk 4]
```diff
@@ -24,13 +24,16 @@ import (
 	"time"
 
 	appspub "github.com/openkruise/kruise/apis/apps/pub"
+	"github.com/openkruise/kruise/pkg/features"
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_test.go:27]: plz add ut for defaultPatchUpdateSpecToPod

File: pkg/util/inplaceupdate/inplace_update_defaults.go (hunk 2 above)
Comment (FillZpp) [reply pkg/util/inplaceupdate/inplace_update_defaults.go:266]: There is not only the iteration, but also calculate strategic patch using old and new templates.

File: pkg/util/inplaceupdate/inplace_update_defaults.go (hunk 3 above)
Comment (FillZpp) [reply pkg/util/inplaceupdate/inplace_update_defaults.go:281]: done

File: pkg/util/inplaceupdate/inplace_update_test.go (hunk 4 above)
Comment (FillZpp) [reply pkg/util/inplaceupdate/inplace_update_test.go:27]: done

**Review by furykerry — COMMENTED** [review]
/lgtm

**Pull request merged**